File: app/src/main/java/com/tgk/app/core/PinyinInterpreter.kt
package com.tgk.app.core

class PinyinInterpreter(private val output: (String) -> Unit) {
    private val vars = mutableMapOf<String, String>()

    fun run(code: String) {
        code.lines().forEach { line ->
            val trimmed = line.trim()
            when {
                trimmed.startsWith("dayin") -> handlePrint(trimmed)
                trimmed.startsWith("bianliang") -> handleVar(trimmed)
                trimmed.startsWith("ruguo") -> output("条件暂不支持")
            }
        }
    }

    private fun handlePrint(line: String) {
        val regex = """dayin\s*\(\s*"(.*?)"\s*\)""".toRegex()
        val match = regex.find(line)
        match?.let {
            var content = it.groupValues[1]
            vars.forEach { (k, v) -> content = content.replace("$$k", v) }
            output(content)
        }
    }

    private fun handleVar(line: String) {
        val regex = """bianliang\s+(\w+)\s*=\s*(.*);""".toRegex()
        val match = regex.find(line)
        match?.let {
            vars[it.groupValues[1]] = it.groupValues[2].trim('"')
        }
    }
}

File: MauiUI/Core/PinyinInterpreter.cs
using System.Text.RegularExpressions;
namespace TianGongKaiWu.Core;
public class PinyinInterpreter
{
    private Dictionary<string, object> vars = new();
    private Dictionary<string, string> grammar;
    public event Action<string> OnOutput;
    public PinyinInterpreter(Dictionary<string, string> grammarDict)
    {
        grammar = grammarDict;
    }
    public void Run(string code)
    {
        var lines = code.Split('\n');
        foreach (var line in lines)
        {
            ExecuteLine(line.Trim());
        }
    }
    private void ExecuteLine(string line)
    {
        if (string.IsNullOrEmpty(line)) return;
        if (line.StartsWith("dayin")) HandlePrint(line);
        else if (line.StartsWith("bianliang")) HandleVar(line);
        else if (line.StartsWith("ruguo")) OnOutput?.Invoke("(条件语句暂未实现，请使用盘古转换)");
    }
    private void HandlePrint(string line)
    {
        var match = Regex.Match(line, @"dayin\s*\(\s*""?(.*?)""?\s*\)");
        if (match.Success)
        {
            string content = match.Groups[1].Value;
            foreach (var v in vars)
                content = content.Replace($"${v.Key}", v.Value.ToString());
            OnOutput?.Invoke(content);
        }
    }
    private void HandleVar(string line)
    {
        var match = Regex.Match(line, @"bianliang\s+(\w+)\s*=\s*(.*);");
        if (match.Success)
        {
            string name = match.Groups[1].Value;
            string value = match.Groups[2].Value.Trim('"');
            vars[name] = value;
        }
    }
}
